packages/flutter/lib/src/services/autofill.dart

```diff
@@ -0,0 +1,273 @@
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'uniqueIdentifier': uniqueIdentifier,
+      'hints': autofillHints,
+      'editingValue': currentEditingValue.toJSON(),
+    };
+  }
+}
+
+/// A client that represents an autofillable input field in the autofill workflow.
+abstract class AutofillClient {
+  /// The unique identifier of this [AutofillClient].
+  ///
+  /// This [AutofillClient] will not participant in autofill until the [autofillId]
+  /// becomes non-null.
+  String get autofillId;
+
+  /// The [TextInputConfiguration] that describes this [AutofillClient].
+  TextInputConfiguration get textInputConfiguration;
+
+  /// Requests this [AutofillClient] update its [TextEditingState] to the given
+  /// state.
+  void updateEditingValue(TextEditingValue newEditingValue);
+}
+
+/// A [TextInputClient] that triggers autofill when focused.
+///
+/// This is typically a [State] or [Element], and should be able to gain focus in
+/// order to trigger autofill.
+abstract class AutofillTrigger extends TextInputClient {
+
+  /// The [AutofillScope] that this [AutofillTrigger] belongs to.
+  ///
+  /// This [AutofillTrigger] will participant in autofill alone if null.
+  AutofillScope get currentAutofillScope;
+}
+
+mixin AutofillClientMixin implements AutofillClient {
+  @override
+  String get autofillId {
+    final AutofillConfiguration configuration = textInputConfiguration.autofillConfiguration;
+    final String identifier = configuration?.uniqueIdentifier;
+    assert(configuration == null || identifier != null);
+    return identifier;
+  }
+}
+
+/// An ordered group of [AutofillClient]s that are logically connected.
+///
+/// {@template flutter.services.autofill.AutofillScope}
+/// [AutofillClient]s within the same [AutofillScope] are isolated from other
+/// input fields during autofill. That is, when an [AutofillTrigger] gains focus,
+/// only the [AutofillClient]s within the same [AutofillScope] will be visible to
+/// the autofill service, in the same order as they appear in [autofillClients].
+///
+/// [AutofillScope] also allows [TextInput] to redirect autofill values from the
+/// platform to the [AutofillClient] with the given identifier, by calling
+/// [getAutofillClient].
+///
+/// An [AutofillClient] not tied to any [AutofillScope] will only participant in
+/// autofill if the autofill is directly triggered by its own [AutofillTrigger].
+/// {@endtemplate}
+mixin AutofillScope {
+  /// Gets the [AutofillScope] associated with the given [uniqueIdentifier], in
+  /// this [AutofillScope].
+  ///
+  /// Returns null if there's no matching [AutofillClient].
+  AutofillClient getAutofillClient(String uniqueIdentifier);
+
+  /// The collection of [AutofillClient]s currently tied to this [AutofillScope].
+  ///
+  /// The [AutofillClient]s should appear in a sensible order, as the autofill
+  /// service will see these [AutofillClient]s in the exact same order.
```

Comment:
For example with the following form on iOS:
1. email
2. username
3. password

username and password will be autofilled if the user autofills the password field. 
But if the email field and the username field swap places, it will be the email field and the password field that get autofilled if the user does the same.